smithy-build/src/main/java/software/amazon/smithy/build/transforms/FlattenNamespaces.java

```diff
@@ -0,0 +1,181 @@
+        }
+
+        /**
+         * @return Gets the service shape ID of the service that will have
+         * its shape namespaces updated.
+         */
+        public ShapeId getService() {
+            return service;
+        }
+
+        /**
+         * Sets the set of tags that are retained in the model.
+         *
+         * @param tags The tags to retain in the model.
+         */
+        public void setIncludeTagged(Set<String> tags) {
+            this.tags = tags;
+        }
+
+        /**
+         * Gets the set of tags that are retained in the model.
+         *
+         * @return Returns the tags to retain.
+         */
+        public Set<String> getIncludeTagged() {
+            return tags;
+        }
+    }
+
+    @Override
+    public Class<Config> getConfigType() {
+        return Config.class;
+    }
+
+    @Override
+    protected Model transformWithConfig(TransformContext context, Config config) {
+        if (config.getService() == null || config.getNamespace() == null) {
+            throw new SmithyBuildException(
+                    "'namespace' and 'service'properties must be set on flattenNamespace transformer.");
+        }
+        Model model = context.getModel();
+        Map<ShapeId, ShapeId> shapesToRename = getRenamedShapes(config, model);
+        return ModelTransformer.create().renameShapes(model, shapesToRename);
+    }
+
+    @Override
+    public String getName() {
+        return "flattenNamespaces";
+    }
+
+    private Map<ShapeId, ShapeId> getRenamedShapes(Config config, Model model) {
+        if (!model.getShape(config.getService()).isPresent()) {
+            throw new SmithyBuildException("Service not found in model when performing flattenNamespaces transform.");
+        }
+
+        Map<ShapeId, ShapeId> shapesToRename = getRenamedShapesConnectedToService(config, model);
+        Set<ShapeId> taggedShapesToInclude = getTaggedShapesToInclude(config.getIncludeTagged(), model);
+
+        for (ShapeId id : taggedShapesToInclude) {
+            ShapeId updatedShapeId = updateNamespace(id, config.getNamespace());
+            // If new shape ID already exists in map of shapes to rename, skip
+            // including the additional shape to avoid a conflict.
+            if (!shapesToRename.containsValue(updatedShapeId)) {
+                shapesToRename.put(id, updatedShapeId);
+            }
+        }
+
+        return shapesToRename;
+    }
+
+    private ShapeId updateNamespace(ShapeId shapeId, String namespace) {
+        if (shapeId.getMember().isPresent()) {
+            return ShapeId.fromParts(namespace, shapeId.getName(), shapeId.getMember().get());
+        }
+        return ShapeId.fromParts(namespace, shapeId.getName());
+    }
+
+    private Map<ShapeId, ShapeId> getRenamedShapesConnectedToService(Config config, Model model) {
+        Walker shapeWalker = new Walker(model.getKnowledge(NeighborProviderIndex.class, NeighborProviderIndex::new)
```

Comment:
`model.getKnowledge(NeighborProviderIndex.class, NeighborProviderIndex::new)` should be `NeighborProviderIndex.of(model)`.